### ISSUE TITLE
QUIC: Peer closing connection can result in OperationCanceledException on pending write

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/runtime#58078]
I found this while looking into test issues.

Here's what I see:
(a) Server tries to send a GOAWAY on the client's control stream; this write pends
(b) Client closes the connection (this races with (a), but it happens pretty consistently)
(c) Server receives PEER_SEND_COMPLETE from msquic, and completes the pending write from (a) with OperationCanceledException: "Write was canceled"

I would expect this to result in QuicConnectionAbortedException instead.

Stack trace for (c) above:

```
  !!!!! ShutdownAsync: got exception System.OperationCanceledException: Write was canceled
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.HandleEventSendComplete(State state, StreamEvent& evt) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 1139
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.HandleEvent(State state, StreamEvent& evt) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 854
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.NativeCallbackHandler(IntPtr stream, IntPtr context, StreamEvent& streamEvent) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 830
  --- End of stack trace from previous location ---
     at System.Net.Quic.Implementations.MsQuic.Internal.ResettableCompletionSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\Internal\ResettableCompletionSource.cs:line 73
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.WriteAsync(ReadOnlyMemory`1 buffer, Boolean endStream, CancellationToken cancellationToken) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 296
     at System.Net.Test.Common.Http3LoopbackStream.SendFrameHeaderAsync(Int64 frameType, Int32 payloadLength) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 162
     at System.Net.Test.Common.Http3LoopbackStream.SendFrameAsync(Int64 frameType, ReadOnlyMemory`1 framePayload) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 167
     at System.Net.Test.Common.Http3LoopbackStream.SendGoAwayFrameAsync(Int64 firstInvalidStreamId) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 150
     at System.Net.Test.Common.Http3LoopbackConnection.ShutdownAsync() in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackConnection.cs:line 249
      System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_HttpClientHandlerTest_Headers_Http3_MsQuic.GetAsync_EmptyResponseHeader_Success [FAIL]
```

**Comment by ManickaP** [comment]
AFAIU it's because the event itself contains canceled flag:
https://github.com/dotnet/runtime/blob/bc7081e81dc34ebe3efa44829e6396f94ac01b38/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs#L521-L521

The choice of the exception here is up to us, so I assume we can change it without any problems.

Triage: we should revisit this in 7.0 frame, related to #55619

cc @CarnaViire

**Comment by CarnaViire** [comment]
I've seen exactly same thing happening while investigating https://github.com/dotnet/runtime/issues/53090.

The cause of OperationCanceledException is indeed `Canceled` flag inside PEER_SEND_COMPLETE event. It is not as easy to mitigate though.

1) We cannot just change OperationCanceledException to QuicConnectionAbortedException, because:
    - Canceled=true might be set due to another reason, e.g. peer aborting reading (without closing the connection)
    - Even if it was canceled due to connection being closed, we don't have Connection's error code at that moment, because related events didn't happen yet

2) I though I could just ignore PEER_SEND_COMPLETE if Canceled=true and defer handling to the following aborting events (PEER_RECEIVE_ABORTED or SHUTDOWN_COMPLETE with ConnectionClosed=true) but I've started getting INVALID_STATE exception from StreamSendDelegate instead. I think it's not *because* I ignored the event, but it's just another type of possible outcome from the race between writing and closing connection, which just got more visible. I think so because when ignoring the event, I left SendState in Pending, and if the next Write would start executing for some reason, it would fail on Debug.Assert in HandleWriteStartState.
